### PR TITLE
Split `Vec::dedup_by` into 2 cycles

### LINKED ISSUE #77772: Potentially faster dedup_by implementation
It seems like the `dedup_by` implementation makes two passes over the data, making many more copies than are strictly necessary. Doing this allows the `same_bucket` implementation to panic without causing a leak. I wouldn't call this *bad*, but it was a surprise to see that it was doing the extra work.

https://godbolt.org/z/4To1a1

Mimicking the way `Drain` handles this and temporarily shortening the vector allows us to cut down on the copies (and also conveniently removes a panicking branch).

I wonder if it makes sense to document the trade-off that `std` makes here? Imo, it'd make it easier to use the library correctly in the way that <https://doc.rust-lang.org/std/vec/struct.Vec.html#current-implementation-7> does.

### PR DESCRIPTION
**Pull request by AngelicosPhosphoros**
First cycle runs until we found 2 same elements, second runs after if there any found in the first one. This allows to avoid any memory writes until we found an item which we want to remove.

This leads to significant performance gains if all `Vec` items are kept: -40% on my benchmark with unique integers.

Results of benchmarks before implementation (including new benchmark where nothing needs to be removed):
 *   vec::bench_dedup_all_100                 74.00ns/iter  +/- 13.00ns
 *   vec::bench_dedup_all_1000               572.00ns/iter +/- 272.00ns
 *   vec::bench_dedup_all_100000              64.42µs/iter  +/- 19.47µs
 *   __vec::bench_dedup_none_100                67.00ns/iter  +/- 17.00ns__
 *   __vec::bench_dedup_none_1000              662.00ns/iter  +/- 86.00ns__
 *   __vec::bench_dedup_none_10000               9.16µs/iter   +/- 2.71µs__
 *   __vec::bench_dedup_none_100000             91.25µs/iter   +/- 1.82µs__
 *   vec::bench_dedup_random_100             105.00ns/iter  +/- 11.00ns
 *   vec::bench_dedup_random_1000            781.00ns/iter  +/- 10.00ns
 *   vec::bench_dedup_random_10000             9.00µs/iter   +/- 5.62µs
 *   vec::bench_dedup_random_100000          449.81µs/iter  +/- 74.99µs
 *   vec::bench_dedup_slice_truncate_100     105.00ns/iter  +/- 16.00ns
 *   vec::bench_dedup_slice_truncate_1000      2.65µs/iter +/- 481.00ns
 *   vec::bench_dedup_slice_truncate_10000    18.33µs/iter   +/- 5.23µs
 *   vec::bench_dedup_slice_truncate_100000  501.12µs/iter  +/- 46.97µs

Results after implementation:
 *   vec::bench_dedup_all_100                 75.00ns/iter   +/- 9.00ns
 *   vec::bench_dedup_all_1000               494.00ns/iter +/- 117.00ns
 *   vec::bench_dedup_all_100000              58.13µs/iter   +/- 8.78µs
 *   __vec::bench_dedup_none_100                52.00ns/iter  +/- 22.00ns__
 *   __vec::bench_dedup_none_1000              417.00ns/iter +/- 116.00ns__
 *   __vec::bench_dedup_none_10000               4.11µs/iter +/- 546.00ns__
 *   __vec::bench_dedup_none_100000             40.47µs/iter   +/- 5.36µs__
 *   vec::bench_dedup_random_100              77.00ns/iter  +/- 15.00ns
 *   vec::bench_dedup_random_1000            681.00ns/iter  +/- 86.00ns
 *   vec::bench_dedup_random_10000            11.66µs/iter   +/- 2.22µs
 *   vec::bench_dedup_random_100000          469.35µs/iter  +/- 20.53µs
 *   vec::bench_dedup_slice_truncate_100     100.00ns/iter   +/- 5.00ns
 *   vec::bench_dedup_slice_truncate_1000      2.55µs/iter +/- 224.00ns
 *   vec::bench_dedup_slice_truncate_10000    18.95µs/iter   +/- 2.59µs
 *   vec::bench_dedup_slice_truncate_100000  492.85µs/iter  +/- 72.84µs

Resolves #77772

P.S. Note that this is same PR as #92104 I just missed review then forgot about it.
Also, I cannot reopen that pull request so I am creating a new one.
I responded to remaining questions directly by adding commentaries to my code.

### CODE REVIEW

**Comment by AngelicosPhosphoros** [comment]
@joshtriplett Hi, I just want to notify that there is an unreviewed PR.

**Comment by the8472** [comment]
r? the8472

File: library/alloc/benches/vec.rs [hunk 1]
```diff
@@ -658,73 +658,160 @@ fn random_sorted_fill(mut seed: u32, buf: &mut [u32]) {
     buf.sort();
 }
 
-fn bench_vec_dedup_old(b: &mut Bencher, sz: usize) {
+// Measures performance of slice dedup impl.
+// "Old" implementation of Vec::dedup
```
Comment (the8472) [inline library/alloc/benches/vec.rs:662]: I think this is a bit misleading in the context of this PR. It's "old-old" now. Can we put a version on it "from Rust 1.xx" or something like that?

File: library/alloc/src/vec/mod.rs [hunk 2]
```diff
@@ -1775,7 +1775,31 @@ impl<T, A: Allocator> Vec<T, A> {
             return;
         }
 
-        /* INVARIANT: vec.len() > read >= write > write-1 >= 0 */
+        // Check if we ever want to remove anything.
+        // This allows to use copy_non_overlapping in next cycle.
+        // And avoids any memory writes if we don't need to remove anything.
+        let mut possible_remove_idx = 1;
```
Comment (the8472) [inline library/alloc/src/vec/mod.rs:1781]: The name is a bit weird. Maybe `first_duplicate`?

File: library/alloc/src/vec/mod.rs [hunk 3]
```diff
@@ -1775,7 +1775,31 @@ impl<T, A: Allocator> Vec<T, A> {
...
+            };
+            if found_duplicate {
+                break;
+            }
+            possible_remove_idx += 1;
+        }
+        // Don't need to remove anything.
+        if possible_remove_idx >= len {
+            return;
+        }
```
Comment (the8472) [inline library/alloc/src/vec/mod.rs:1800]: If it can exceed the length then that's problematic for `Vec<()>` since that can have a length of `usize::MAX`.

Comment (AngelicosPhosphoros) [reply library/alloc/src/vec/mod.rs:1800]: Well, valid indexes are always smaller than length so even in that case there would not be overflow.

Comment (the8472) [reply library/alloc/src/vec/mod.rs:1800]: So the `>` in `>=` isn't really necessary? Ok.

**Comment by AngelicosPhosphoros** [comment]
@the8472 I made changes you requested.

**Pull request merged**